types/field_type.go

```diff
@@ -41,12 +41,7 @@ func NewFieldType(tp byte) *FieldType {
 		Flen:    UnspecifiedLength,
 		Decimal: UnspecifiedLength,
 	}
-	if tp != mysql.TypeVarchar && tp != mysql.TypeVarString && tp != mysql.TypeString {
-		ft.Collate = charset.CollationBin
-	} else {
-		ft.Collate = mysql.DefaultCollationName
-	}
-	// TODO: use DefaultCharsetForType to set charset and collate
+	ft.Charset, ft.Collate = DefaultCharsetForType(tp)
```

Comment:
the collate size is equivalent, but we now change `Charset` from `nil` to `DefaultCharset` and update tests correspondingly. Can you elaborate the background and how it won't affect caller assumption or user facing behavior?